docs/configuration.md
# 配置项

全部配置项都在 config.py 中可以找到。但是从3.0版本开始不再使用`local_settings.py`存放配置，统一使用`config.yaml`，具体可以看`config.yaml.tmpl`样板文件中的设置项

## DB_URL

- 默认值：`mysql://root:@localhost:3306/test?charset=utf8`

设置数据库的URL，另外也支持从环境变量读取

## DEBUG

- 默认值：`False`

是否开启DEBUG模式，开启后可以看到详细错误，修改代码后不用重启就可以autoreload。另外也支持从环境变量读取

## REDIS_URL

- 默认值: `redis://localhost:6379`

Redis服务器地址，另外也支持从环境变量读取

## WTF_CSRF_SECRET_KEY

- 默认值：`123`

Sanic-wtf 需要的 CSRF KEY

## MEMCACHED_HOST

- 默认值: `127.0.0.1`

Memcached的主机名或者IP，另外也支持从环境变量读取

## MEMCACHED_PORT

- 默认值: `11211`

Memcached的端口

## redirect_uri

- 默认值: `http://127.0.0.1:8000/oauth`

在Github申请的应用的回调地址，如果只是使用默认的 http://127.0.0.1:8000 不需要改变

## client_id

- 默认值: `098a2e6da880878e05da`

Github申请的应用client_id，如果只是使用默认的 http://127.0.0.1:8000 不需要改变

## client_secret

- 默认值: `854cc0d86e61a83bb1dd00c3b23a3cc5b832d45c`

Github申请的应用client_secret，如果只是使用默认的 http://127.0.0.1:8000 不需要改变

## REACT_PROMPT

- 默认值: `喜欢这篇文章吗? 记得给我留言或订阅哦`

文章下的默认提示语

## SHOW_PROFILE

- 默认值: `False`

是否显示个人设置项，开启后需要通过后台配置

注意: 这会改变主要的布局！

## OWNER

- 默认值: `xiaoming`

指定博客作者名字

## SITE_TITLE

- 默认值: `My Blog`

指定博客标题

## PER_PAGE

- 默认值: `10`

首页文章列表每页文章数量

## GOOGLE_ANALYTICS

- 默认值: `''`

指定GA，为空表示不开启

## SENTRY_DSN

- 默认值: `''`

指定Sentry的DSN(Data Source Name)，为空表示不追踪错误堆栈

## SITE_NAV_MENUS

可以定制导航栏，每项参数最少2个(Endpoint, Name)，最多5个(Endpoint, Name, FontAwesome icon, 颜色)，具体默认值可以看 config.py

## BEIAN_ID

可以设置ICP备案号

## JWT_SECRET

- 默认值: `lyanna`

JWT的秘钥

## EXPIRATION_DELTA

- 默认值: `3600(秒)`

JWT的超时时间

## MAIL_SERVER

- 默认值: `smtp.qq.com`

用来发邮件的SMTP服务器主机，为空将不会发邮件

## MAIL_PORT

- 默认值: `465`

用来发邮件的SMTP服务端口

## MAIL_USERNAME

- 默认值: `''`

用来发邮件的账号，为空将不会发邮件

## MAIL_PASSWORD

- 默认值: `''`

用来发邮件的账号密码，为空将不会发邮件

## BLOG_URL

- 默认值: `https://example.com`

博客的地址，在提及邮件中可以直接访问博客对应文章页面

## SHOW_PAGEVIEW

- 默认值: `False`

是否显示文章阅读量(PV)，建议有一定访问量后再显示出来

## PERMALINK_TYPE

- 默认值: `slug`

文章页面URL的类型。可选值: id、slug、title。举个例子, PERMALINK_TYPE为id时，首页、分类页、标签页等页面都会使用`https://example.com/post/{id}`这样效果的URL，帮助SEO。

## REDIS_SENTINEL_SERVICE_HOST

- 默认值: `None`

除了用上面提到的`REDIS_URL`指定Redis服务器，还可以用分布式的Redis Sentinel集群架构，本项指定主机地址，默认不开启。另外也支持从环境变量读取

## REDIS_SENTINEL_SERVICE_PORT

- 默认值: `26379`

Redis Sentinel集群架构，本项指定主机端口号。另外也支持从环境变量读取

## SHOW_AUTHOR

- 默认值: `False`

文章页可以显示发布文章的用户信息(名字、头像)，目前还不提供链接到用户详情页

## USE_FFMPEG

动态可以上传视频，如果服务器上安装了ffmpeg可以在上传后自动截取封面并且获得视频大小，这样让用户访问体验更好。本项会自动根据是否安装ffmpeg决定，为了在启动时给博主提示，改变此项无意义

## CDN_DOMAIN

- 默认值: `''`

为了更好地访问效果，博客内的静态文件(Javascript/css/img/fonts等)地址可以使用CDN的域名，默认不开启

## LIMIT_RSS_CRAWLING

- 默认值: `False`

开启后，RSS只显示全文的80%。注意，不显示全问会影响了订阅用户，但能提高搜索权重

## READ_MORE

- 默认值: `<p>原文: <a href="{url}">{title}</a></p>`

在RSS内容中添加原文链接，文案支持HTML，支持2个内置参数: `{url}`表示文章链接，`{title}`表示文章标题。

## SYNC_DOUBAN_ACTIVITIES

- 默认值: `''`

默认不开启，可以设置自己的`DOUBAN_ID`(https://www.douban.com/people/:id/)。这样当你在豆瓣上标记了电影/读书/游戏后，会定期同步到博客上

## SYNC_INTERVAL

- 默认值: `120`

从豆瓣同步动态的间隔，默认每2小时同步一次(单位: 分钟)。


docs/changelog.md
# Changelog

## 4.0 (2022-05-23)

1. Add note type
2. Show my favorite subjects(game, book, movie)
3. Add more widgets(latest_notes, games, books, movies), with full page
4. Post page can show subject card(movie, book, game)
5. Compatible with Python 3.10
6. Use Vue3 & Element-Plus
7. Sync your douban activities(movie, book, game)

## 3.5.1 (2021-09-02)

1. Fix `ident` params

## 3.5 (2021-08-27)

1. Bump the version of the dependent libraries to the latest
2. Delete the corresponding activity after deleting the post
3. Fix topic page params
4. Fix created_at format

## 3.4 (2020-11-01)

1. Add LIMIT_RSS_CRAWLING & READ_MORE
2. Disable specify UA and empty UA access

## 3.3 (2020-01-19)

1. Use aioredis instead of aiomcache
2. Record response spent time
3. Optimize pagination

## 3.2 (2020-01-10)

1. Optimize dist files size
2. Fix docker env

## 3.1 (2020-01-07)

1. Fix some issues
2. Add task log
3. Fix aio-libs/aiomysql#454

## 3.0.1 (2019-12-24)

1. @cache support `serialize` & `parser` opts
2. Fix some issues
3. Add activity responsive style

## 3.0 (2019-12-22)

1. Use Python 3.8 syntax
2. Activity feature. like douban user's status
3. Configuration file uses YAML format
4. Use isort/bandit/autopep8
5. Use mypy
6. Support reply comment

## 2.7.1 (2019-11-27)

1. Update recent comments in realtime
2. Fix github user save
3. Some minor modifications

## 2.7 (2019-11-26)

1. Sidebar Widgets(with aboutme、blogroll、most\_viewed、latest\_comments、tagcloud、html)
2. Mention autocomplete
3. Keep the order of the tags in the article
4. Remove profile feature
5. Use iconfont
6. Add Special topic feature
7. Custom menu icon & color
8. Docker compose support
9. Kubernetes support

## 2.6.1 (2019-09-17)

1. Text spacing for link
2. Fix tag cache

## 2.6 (2019-08-31)

1. Support github-cards
2. Fix comment's reflected xss vulnerability
3. Use Pangu
4. Fix some issues

## 2.5 (2019-08-20)

1. Fix some issues
2. Support permalink types
3. Show pageview
4. Optimize bundle files
5. Use simple marked editor

## 2.4 (2019-05-30)

1. Fix some issues
2. Perf: use asyncio.gather

## 2.3 (2019-04-19)

1. Support @mentions, and send an email
2. Profile use attrs
3. Use a better way to fix `OverflowError`
4. Use the user avatar set by the user, otherwise use the default avatar
5. Discard `pipenv` installation method

## 2.2 (2019-04-03)

1. Fix admin post api
2. Filter unpiblished posts
3. Fix `get_all`

## 2.1 (2019-03-21)

1. Fix some issues
2. Use my aiomcache's fork

## 2.0 (2019-03-09)

Bump 2.0

## 1.1 (2019-01-20)

Bump 1.1

## 1.0 (2019-01-17)

Init commit


config.py
import distutils.spawn
import os
from pathlib import Path
from typing import List, Tuple

import yaml


class AttrDict(dict):

    def __init__(self, *args, **kwargs) -> None:
        super(AttrDict, self).__init__(*args, **kwargs)
        self.__dict__ = self


DB_URL = os.getenv('DB_URL', 'mysql://root:@localhost:3306/test?charset=utf8mb4')  # noqa
REDIS_URL = os.getenv('REDIS_URL', 'redis://localhost:6379')
DEBUG = os.getenv('DEBUG', '').lower() in ('true', 'y', 'yes', '1')
WTF_CSRF_SECRET_KEY = 123
AUTH_LOGIN_ENDPOINT = 'index.login'
MEMCACHED_HOST = os.getenv('MEMCACHED_HOST', '127.0.0.1')
MEMCACHED_PORT = 11211

OAUTH_REDIRECT_PATH = '/oauth'
REDIRECT_URI = 'http://127.0.0.1:8000/oauth'
CLIENT_ID = "098a2e6da880878e05da"
CLIENT_SECRET = "854cc0d86e61a83bb1dd00c3b23a3cc5b832d45c"

REACT_PROMPT = '喜欢这篇文章吗? 记得给我留言或订阅哦'
HERE = Path(__file__).parent.absolute()
UPLOAD_FOLDER = HERE / 'static/upload'
OWNER = 'xiaoming'
SITE_TITLE = 'My Blog'
PER_PAGE = 10
GOOGLE_ANALYTICS = ''
SENTRY_DSN = ''
REQUEST_TIMEOUT = 15
SHOW_PAGEVIEW = False
PERMALINK_TYPE = 'slug'  # 可选 id、slug、title

# [(Endpoint, Name, IconName, Color), ...]
SITE_NAV_MENUS: List[Tuple] = []
BEIAN_ID = ''
JWT_SECRET = 'lyanna'
EXPIRATION_DELTA = 60 * 60
WTF_CSRF_ENABLED = False

MAIL_SERVER = 'smtp.qq.com'
MAIL_PORT = 465
MAIL_USERNAME = ''
MAIL_PASSWORD = ''
MAKO_STRICT_UNDEFINED = False

BLOG_URL = 'https://example.com'

# Redis sentinel
REDIS_SENTINEL_SERVICE_HOST = None
REDIS_SENTINEL_SERVICE_PORT = 26379

SHOW_AUTHOR = False
COMMENT_REACTIONS = ['heart', 'upvote']

USE_FFMPEG = bool(distutils.spawn.find_executable('ffmpeg'))
CDN_DOMAIN = ''

ACCESS_LOG = True
ENABLE_DEBUG_LOG = False

LIMIT_RSS_CRAWLING = False
READ_MORE = ""

SUPER_ADMIN_MODE = False

try:
    from local_settings import *  # noqa
except ImportError:
    ...

try:
    with open(HERE / 'config.yaml') as f:
        config = AttrDict(yaml.safe_load(f))
    partials = config.partials  # type: ignore
    USE_YAML = True
except FileNotFoundError:
    USE_YAML = False
    partials = []
    config = {}  # type: ignore

if USE_YAML:
    for subconfig in [config.common, config.oauth]:  # type: ignore
        globals().update({k.upper(): v for k, v in subconfig.items()})
    globals().update({f'MAIL_{k.upper()}': v for k, v
                      in config.mail.items()})  # type: ignore

if not USE_FFMPEG:
    print('Warning: Cover can be displayed after installing `ffmpeg`')

redis_sentinel_host = os.getenv('REDIS_SENTINEL_SVC_HOST') or REDIS_SENTINEL_SERVICE_HOST  # noqa
if redis_sentinel_host:
    redis_sentinel_port = os.getenv('REDIS_SENTINEL_SVC_PORT',
                                    REDIS_SENTINEL_SERVICE_PORT)
    from redis.sentinel import Sentinel
    sentinel = Sentinel([(redis_sentinel_host, redis_sentinel_port)],
                        socket_timeout=0.1)
    redis_host, redis_port = sentinel.discover_master('mymaster')
    REDIS_URL = f'redis://{redis_host}:{redis_port}'
